apex/contrib/peer_memory/peer_memory.py

```diff
@@ -0,0 +1,70 @@
+import torch
+import numpy as np
+import peer_memory as pm
+
+class PeerMemoryPool(object):
+
+    def __init__(self, rank, world_size, peer_group_size, static_size, dynamic_size):
+        self.peer_group = rank // peer_group_size
+        self.peer_rank = rank % peer_group_size
+        self.peer_group_size = peer_group_size
+        self.alignment = 256
+        self.static_size = ((static_size + self.alignment - 1) // self.alignment) * self.alignment
+        self.dynamic_size = ((dynamic_size + self.alignment - 1) // self.alignment) * self.alignment
+        # allocate giant pool of device memory
+        self.raw = pm.allocate_raw(self.static_size+self.dynamic_size)
+        # exchange peer pointers with nccl
+        raw_ipc = pm.get_raw_ipc_address(self.raw).cuda()
+        peer_raw_ipcs = [torch.empty_like(raw_ipc) for _ in range(world_size)]
+        torch.distributed.all_gather(peer_raw_ipcs, raw_ipc)
+        peer_raw_ipcs = torch.stack(peer_raw_ipcs).cpu()
+        self.peer_raw = pm.get_raw_peers(peer_raw_ipcs, self.peer_rank, self.raw)
+        self.static_offset = 0
+        self.dynamic_offset = 0
+
+    def __del__(self):
+        pm.free_raw(self.raw)
```

Comment:
`free_raw` also does't seem to be defined

Reply:
The changes in this PR means that --fast_bottleneck now depends on two new extensions: --peer_memory and --nccl_p2p. I did not find a way to specify this dependency in setup.py. To compile --fast_bottleneck, the following switches are necessary: 

python setup.py install --cpp_ext --cuda_ext --fast_bottleneck --peer_memory --nccl_p2p

Reply:
To be clear, --fast_bottleneck will compile without --peer_memory and --nccl_p2p, but it won't run correctly.